pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/kubernetes/KubernetesRuntime.java

```diff
@@ -155,6 +158,9 @@
                       String pythonDependencyRepository,
                       String pythonExtraDependencyRepository,
                       String pulsarDockerImageName,
+                      String javaFunctionDockerImageName,
+                      String pythonFunctionDockerImageName,
+                      String goFunctionDockerImageName,
```

Comment:
This can already be achieved with no code changes by implementing your own https://github.com/apache/pulsar/blob/master/pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/kubernetes/KubernetesManifestCustomizer.java
This class allows you to do anything to the function spec so you can set whatever image you want.

Is there another patch coming with this? The kubernetes runtime only supports one docker image right?. So, If I configured a python image I would only be able to run python functions. I'm not sure this is the right approach and think the design needs to be thought through more. Maybe a more generic structure like a map. That way you could have images for specific runtime version java8, java11, python3.7, python3.8, etc..

Reply:
> Is there another patch coming with this? The kubernetes runtime only supports one docker image right?. So, If I configured a python image I would only be able to run python functions.

Yes, this is what I want to achieve. Currently, the default docker image we use is `apachepulsar/pulsar-all`, but sometimes users only want to use the Python Function, then for Java Function and Go Function, we don’t need to build related dependencies to the image, right?

Reply:
> This class allows you to do anything to the function spec so you can set whatever image you want.

You are right, we can use the `KubernetesManifestCustomizer` interface to implement different language function. And then? How do we load user-defined implementation classes? The most direct way is to rebuild a new image based on `apachepulsar/pulsar-all`. Eg:

```
FROM apachepulsar/pulsar-all:2.5.0
COPY my-custer-jar.jar /pulsar/lib
...
```

Or, another option though would be to do the following:
- get your jar into a k8s volume
- mount that volume into your function worker pod (say in `/pulsar/extraLibs` )
- set `PULSAR_EXTRA_CLASSPATH=/pulsar/extraLibs`

So, I think this is two problems:

- We allow users to configure image for a specific language function(Eg: Python function)
- For the user-defined k8s spec, we should consider how to load the user-implemented tar package into the pod.

Reply:
> We allow users to configure image for a specific language function(Eg: Python function)
I agree it would be nice to have images for specific runtimes.

Are users in charge of building their own function images? Or will the project be publishing images for each function runtime?

If users are expected to create their own images thing could get tricky because currently the pulsar-admin tool is required for functions to run in kubernetes. It's used to download the function binary (https://github.com/apache/pulsar/blob/master/pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/kubernetes/KubernetesRuntime.java#L802). This means every image needs java. Ideally, if you are running python or go functions you should not need a java dependency.

> For the user-defined k8s spec, we should consider how to load the user-implemented tar package into the pod.

If you are writing custom plugins it usually makes sense to build your own image and you can dump the plugin in the `pulsar/lib` directory and it will automatically end up on the class path.


Here are a few possible ways to solve multiple images.

1 - the current way with a flat config
```
javaFunctionDockerImageName
pythonFunctionDockerImageName
goFunctionDockerImageName
```

2 - pass a map of runtime to image
I like a map since it allows for easier modifications in the future. However, it can be a little more configuration.
```
runtime -> {}
java -> {
  image: java-image
}
python -> {
  image: python-image
},
go -> {
  image: go-image
},
```

3 - implement a KubernetesManifestCustomizer for multiple images and make that plugin part of the project. The Customizer class is a configuration `runtimeCustomizerClassName` in the function_worker.yaml. This Customizer can also take configuration.


Options 1 and 2 require this code to be updated (
This code needs to be updated (https://github.com/apache/pulsar/blob/master/pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/kubernetes/KubernetesRuntime.java#L973).

Thoughts?

Reply:
@cckellogg I think the idea is to create different instance images for different language runtime. So we can start making the language runtime pluggable.

Reply:
I agree with the concept and idea. I was trying to brainstorm the configuration approach of flat vs a map for the images. Would any other information be needed for specific image? If not then flat is probably ok.